### DISCUSSION TITLE
m1 / arm64 support

### DISCUSSION THREAD

**Discussion by ankopainting** [sfu-db/connector-x#237] [Ideas]
Hi there,

I am a fan of this library.  I am considering using it on my team, but as we are now getting new m1 laptops the lack of arm support is a dealbreaker.  I'd just like to start a discussion about when it could be supported, what needs to be done, if there is anything the community can do to contribute etc.

We are using it with polars if that helps

thanks!

**Comment by ankopainting** [comment]
p.s. linking related issue https://github.com/sfu-db/connector-x/issues/186

**Reply by wangxiaoying** [reply]
Hi @ankopainting , thanks for open this discussion! The reason we do not support m1 is because we use github action to manage our release and it does not host mac m1 for now (https://github.com/actions/virtual-environments/issues/2187). So currently we need to manually build wheel file on apple m1 like the procedure here: https://github.com/sfu-db/connector-x/issues/186#issuecomment-1013733401

**Reply by ankopainting** [reply]
ah this makes sense, do you know if it would be possible to cross compile instead? 

I'm not too familiar with rust w/ python, but it looks like polars somehow does it? https://github.com/pola-rs/polars/blob/master/.github/workflows/create-py-mac-universal2-release.yaml

**Reply by wangxiaoying** [reply]
Thanks for the suggestion! Didn't know this before, will take a look at it.

**Comment by dovahcrow** [comment]
Blocked by the upstream issue: https://github.com/actions/virtual-environments/issues/2187

**Reply by ankopainting** [reply]
hi dovahcrow, 
thanks for the feedback.  I'm just trying to understand why polars manages to do their builds which have arm support.  It seems like they are cross compiling to me, see their config: https://github.com/pola-rs/polars/blob/master/.github/workflows/create-py-mac-universal2-release.yaml

**Reply by dovahcrow** [reply]
@ankopainting Thanks for finding the solution! We will definitely try it.

**Reply by ankopainting** [reply]
thank you for listening :)

**Reply by wangxiaoying** [reply]
Hi @ankopainting , we have published the wheel file for apple m1 starts from version `0.2.5a2`. Please feel free to give it a try.

**Reply by ankopainting** [reply]
this is awesome! sorry for not seeing this until now.  I am trying to get it working and will report back when i have.  so far i'm getting

partially initialized module 'connectorx' has no attribute 'read_sql' (most likely due to a circular import)

**Reply by ankopainting** [reply]
doh! naming my test file connectorx.py did this!!! can't believe that was the issue! it works great, thank you so much!!!

**Reply by ankopainting** [reply]
uh oh, i think something else is going on - it started breaking again.  will get back if i work it out.

**Reply by dovahcrow** [reply]
> uh oh, i think something else is going on - it started breaking again. will get back if i work it out.

Hi @ankopainting did you figure out the issue?

**Reply by ankopainting** [reply]
yes sorry it was environmental, apologies for not getting back.  the connector-x on m1 is awesome!

**Comment by alexander-beedie** [comment]
FYI: looks like official M1 runner support is getting close...
https://github.com/actions/runner/issues/805#issuecomment-1041968059

**Comment by dovahcrow** [comment]
For anyone who is reading this thread: we added [M1 support](https://pypi.org/project/connectorx/0.2.5a2/#files)!

**Reply by tnightengale** [reply]
@dovahcrow It seems like M1 Support is still only available for 0.2.5, and not 0.3.x?

Any plans to release to both platforms on an ongoing basis?

**Reply by alexander-beedie** [reply]
Curious; haven't had have any issues with Apple Silicon since support was added... Can install and run 0.3.1 with no trouble. What errors/issues are you actually seeing?

**Reply by tnightengale** [reply]
EDIT: My issue is detailed in: https://github.com/sfu-db/connector-x/issues/240

@alexander-beedie Interesting... my issue is actually with Docker on M1. Attempting to install via poetry inside a container, using `poetry add connectorx==0.3.1`, results in:
```
  RuntimeError

  Unable to find installation candidates for connectorx (0.3.1)

  at /usr/local/venv/lib/python3.10/site-packages/poetry/installation/chooser.py:72 in choose_for
       68│ 
       69│             links.append(link)
       70│ 
       71│         if not links:
    →  72│             raise RuntimeError(
       73│                 "Unable to find installation candidates for {}".format(package)
       74│             )
       75│ 
       76│         # Get the best link

  • Installing pandas (1.5.0)
  • Installing psycopg2 (2.9.3)
  • Installing pycryptodome (3.15.0)
  • Installing pytest (7.1.3)
  • Installing smart-open (6.3.0)

Failed to add packages, reverting the pyproject.toml file to its original content.
```

System Details:
```
root@045b6d7d7b88:/code# uname -a
Linux 045b6d7d7b88 5.15.49-linuxkit #1 SMP PREEMPT Tue Sep 13 07:51:32 UTC 2022 aarch64 GNU/Linux
```
```
root@045b6d7d7b88:/code# python --version
Python 3.10.2
```
```
root@045b6d7d7b88:/code# poetry --version
Poetry version 1.1.13
```

**Reply by EarthlyZ9** [reply]
@tnightengale Same issue here :( Did you find any solutions?
